File: HallData.EMS.ApplicationViews/UI/DataViewColumn.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using HallData.ApplicationViews;
using Newtonsoft.Json;
using HallData.Validation;

namespace HallData.EMS.ApplicationViews.UI
{
    public class DataViewColumnKey : IHasKey
    {
        [UpdateOperationParameter]
        [ChildKeyOperationParameter]
        public int? DataViewColumnId { get; set; }
        [JsonIgnore]
        public int Key
        {
            get
            {
                return this.DataViewColumnId ?? 0;
            }
            set
            {
                this.DataViewColumnId = value;
            }
        }
    }
    public class DataViewColumnDisplayName : DataViewColumnKey
    {
        [AddOperationParameter]
        [UpdateOperationParameter]
        [GlobalizedRequired]
        public string DisplayName { get; set; }
    }
    public class DataViewColumnResultName : DataViewColumnDisplayName
    {
        [AddOperationParameter]
        [UpdateOperationParameter]
        [GlobalizedRequired]
        public string ResultName { get; set; }
    }
    public class DataViewColumn<TDataView> : DataViewColumnResultName
        where TDataView: DataViewKey
    {
        [ChildView]
        [GlobalizedRequired]
        public TDataView DataView { get; set; }
        [AddOperationParameter]
        [UpdateOperationParameter]
        public int ResultSetIndex { get; set; }
        [AddOperationParameter]
        [UpdateOperationParameter]
        public bool IsRequired { get; set; }
        [AddOperationParameter]
        [UpdateOperationParameter]
        [GlobalizedRequired]
        public string Name { get; set; }
        
    }
    public class DataViewColumnBase : DataViewColumn<DataViewKey> { }
    public class DataViewColumn : DataViewColumn<DataViewName> { }

}

File: HallData.Admin.Web/Scripts/framework/ui/data/contracts/dataviewcolumn/interfaces/DataViewColumn.ts
module HallData.UI.DataContracts {
    export interface IDataViewColumnKey {
        DataViewColumnId: number;
    }
    export interface IDataViewColumnDisplayName extends IDataViewColumnKey {
        DisplayName: string;
    }
    export interface IDataViewColumnResultName extends IDataViewColumnDisplayName {
        ResultName: string;
    }
    
    export interface IDataViewColumn<TDataView extends IDataViewKey> extends IDataViewColumnResultName {
        IsRequired: boolean;
        Name: string;
        DataView: TDataView;
        ResultSetIndex: number;
    }
    export interface IDataViewColumnBase extends IDataViewColumn<IDataViewKey> { }
    export interface IDataViewColumnResult extends IDataViewColumn<IDataViewName> { }
}